packages/amplify-cli-core/src/feature-flags/featureFlagProvider.ts

```diff
@@ -0,0 +1,422 @@
+    }
+
+    if (flagRegistrationEntry.type !== type) {
+      throw new Error(`'${flagName}' is a ${flagRegistrationEntry.type} type, not ${type}`);
+    }
+
+    let value: T | undefined;
+
+    // Check if effective values has a value for the requested flag
+    if (FeatureFlagProvider.instance.effectiveFlags[parts[0]] && FeatureFlagProvider.instance.effectiveFlags[parts[0]][parts[1]]) {
+      value = <T>FeatureFlagProvider.instance.effectiveFlags[parts[0]][parts[1]];
+    }
+
+    // If there is no value, return the registered defaults for existing projects
+    if (!value) {
+      value = <T>(flagRegistrationEntry.defaultValueForExistingProjects as unknown);
+    }
+
+    return value;
+  };
+
+  private buildJSONSchemaFromRegistrations = () => {
+    return [...this.registrations.entries()].reduce<JSONSchema7>(
+      (s: JSONSchema7, r: [string, FeatureFlagRegistration[]]) => {
+        // r is a tuple, 0=section, 1=array of registrations for that section
+        const currentSection = <JSONSchema7>(s.properties![r[0].toLowerCase()] ?? {
+          type: 'object',
+          additionalProperties: false,
+        });
+
+        currentSection.properties = r[1].reduce<{ [key: string]: JSONSchema7Definition }>((p, fr) => {
+          p![fr.name.toLowerCase()] = {
+            type: fr.type,
+            default: fr.defaultValueForNewProjects,
+          };
+
+          return p;
+        }, {});
+
+        s.properties![r[0].toLowerCase()] = currentSection;
+
+        return s;
+      },
+      {
+        $schema: 'http://json-schema.org/draft-07/schema#',
+        type: 'object',
+        additionalProperties: false,
+        properties: {},
+      },
+    );
+  };
+
+  private buildDefaultValues = (): void => {
+    this.newProjectDefaults = [...this.registrations.entries()].reduce<FeatureFlagsEntry>(
+      (result: FeatureFlagsEntry, r: [string, FeatureFlagRegistration[]]) => {
+        // r is a tuple, 0=section, 1=array of registrations for that section
+        const nest = r[1].reduce<{ [key: string]: {} }>((p, fr) => {
+          p![fr.name] = fr.defaultValueForNewProjects;
+
+          return p;
+        }, {});
+
+        result[r[0]] = {
+          ...result[r[0]],
+          ...nest,
+        };
+
+        return result;
+      },
+      {},
+    );
+
+    this.existingProjectDefaults = [...this.registrations.entries()].reduce<FeatureFlagsEntry>(
+      (result: FeatureFlagsEntry, r: [string, FeatureFlagRegistration[]]) => {
+        // r is a tuple, 0=section, 1=array of registrations for that section
+        const nest = r[1].reduce<{ [key: string]: {} }>((p, fr) => {
+          p![fr.name] = fr.defaultValueForExistingProjects;
+
+          return p;
+        }, {});
+
+        result[r[0]] = {
+          ...result[r[0]],
+          ...nest,
+        };
+
+        return result;
+      },
+      {},
+    );
+  };
+
+  private validateFlags = (allFlags: { name: string; flags: FeatureFlags }[]): void => {
```

Comment:
I'd break out this method and transformEnvFlags into another class / file as well